File: webapp/src/main/java/com/box/l10n/mojito/entity/PollableTask.java
package com.box.l10n.mojito.entity;

import com.box.l10n.mojito.aspect.JsonRawString;
import com.box.l10n.mojito.entity.security.user.User;
import com.box.l10n.mojito.rest.View;
import com.box.l10n.mojito.service.pollableTask.InjectCurrentTask;
import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonRawValue;
import com.fasterxml.jackson.annotation.JsonView;
import java.time.ZonedDateTime;
import java.util.Set;
import javax.persistence.Basic;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.ForeignKey;
import javax.persistence.Index;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OrderBy;
import javax.persistence.Table;
import org.hibernate.annotations.BatchSize;
import org.springframework.data.annotation.CreatedBy;

/** @author jaurambault */
@JsonIgnoreProperties(ignoreUnknown = true)
@Entity
@Table(
    name = "pollable_task",
    indexes = {
      @Index(name = "I__POLLABLE_TASK__NAME", columnList = "name"),
      @Index(name = "I__POLLABLE_TASK__FINISHED_DATE", columnList = "finished_date")
    })
@BatchSize(size = 1000)
public class PollableTask extends AuditableEntity {

  /**
   * Constant that can be passed to functions that use {@link InjectCurrentTask} for readability
   * purpose.
   */
  public static final PollableTask INJECT_CURRENT_TASK = null;

  @Basic(optional = false)
  @Column(name = "name")
  private String name;

  @Column(name = "finished_date")
  private ZonedDateTime finishedDate;

  @JsonIgnore
  @Column(name = "message", length = Integer.MAX_VALUE)
  private String message;

  @JsonIgnore
  @Column(name = "error_message", length = Integer.MAX_VALUE)
  private String errorMessage;

  @Column(name = "error_stacks", length = Integer.MAX_VALUE)
  private String errorStack;

  @Basic(optional = false)
  @Column(name = "expected_sub_task_number")
  private int expectedSubTaskNumber = 0;

  @OneToMany(mappedBy = "parentTask", fetch = FetchType.EAGER)
  @OrderBy("id")
  @BatchSize(size = 1000)
  private Set<PollableTask> subTasks;

  @JsonBackReference
  @ManyToOne
  @JoinColumn(
      name = "parent_task_id",
      foreignKey = @ForeignKey(name = "FK__POLLABLE_TASK__POLLABLE_TASK__ID"))
  private PollableTask parentTask;

  @JsonIgnore
  @Column(name = "timeout")
  private Long timeout;

  @CreatedBy
  @ManyToOne
  @JoinColumn(
      name = BaseEntity.CreatedByUserColumnName,
      foreignKey = @ForeignKey(name = "FK__POLLABLE_TASK__USER__ID"))
  protected User createdByUser;

  public User getCreatedByUser() {
    return createdByUser;
  }

  public void setCreatedByUser(User createdByUser) {
    this.createdByUser = createdByUser;
  }

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public ZonedDateTime getFinishedDate() {
    return finishedDate;
  }

  public void setFinishedDate(ZonedDateTime finishedDate) {
    this.finishedDate = finishedDate;
  }

  public String getMessage() {
    return message;
  }

  public void setMessage(String message) {
    this.message = message;
  }

  public String getErrorMessage() {
    return errorMessage;
  }

  public void setErrorMessage(String errorMessage) {
    this.errorMessage = errorMessage;
  }

  public String getErrorStack() {
    return errorStack;
  }

  public void setErrorStack(String errorStack) {
    this.errorStack = errorStack;
  }

  public int getExpectedSubTaskNumber() {
    return expectedSubTaskNumber;
  }

  public void setExpectedSubTaskNumber(int expectedSubTaskNumber) {
    this.expectedSubTaskNumber = expectedSubTaskNumber;
  }

  public Set<PollableTask> getSubTasks() {
    return subTasks;
  }

  public void setSubTasks(Set<PollableTask> subTasks) {
    this.subTasks = subTasks;
  }

  public PollableTask getParentTask() {
    return parentTask;
  }

  public void setParentTask(PollableTask parentTask) {
    this.parentTask = parentTask;
  }

  public Long getTimeout() {
    return timeout;
  }

  public void setTimeout(Long timeout) {
    this.timeout = timeout;
  }

  @JsonProperty(value = "message")
  @JsonRawValue
  @JsonRawString
  public String getMessageAsJson() {
    return message;
  }

  @JsonProperty(value = "errorMessage")
  @JsonRawValue
  @JsonRawString
  public String getErrorMessageAsJson() {
    return errorMessage;
  }

  /**
   * NOTE review, don't like to have business logic here... but overkill to move somewhere else
   *
   * <p>Indicates if this task and all its sub tasks are finished.
   *
   * @return {@code true} if this task and all its subtasks are finished else {@code false}
   */
  @JsonProperty
  @JsonView(View.PollableSummary.class)
  public boolean isAllFinished() {

    boolean currentTaskFinished = (getFinishedDate() != null);
    boolean allSubtasksFinished = true;

    // If parent task is finished, check all its subtasks
    if (currentTaskFinished) {

      if (getSubTasks() != null) {
        int numSubtasks = getSubTasks().size();

        // if not all expected subtasks have run => not finished
        if (numSubtasks < getExpectedSubTaskNumber()) {
          allSubtasksFinished = false;
        } else {
          // a task is finished only if all its subtasks are finished
          for (PollableTask pollableTask : subTasks) {
            if (!pollableTask.isAllFinished()) {
              allSubtasksFinished = false;
              break;
            }
          }
        }
      }
    }

    return currentTaskFinished && allSubtasksFinished;
  }
}

File: webapp/src/main/resources/public/js/sdk/entity/PollableTask.js
import User from "./User.js";

export default class PollableTask {
    constructor() {

        /** @type {Number} */
        this.id = 0;

        /** @type {String} */
        this.name = "";

        /** @type {Date} */
        this.finishedDate = null;

        /** @type {String} */
        this.message = "";

        /** @type {String} */
        this.errorMessage = "";

        /** @type {String} */
        this.errorStack = "";

        /** @type {Number} */
        this.expectedSubTaskNumber = 0;

        /** @type {PollableTask[]} */
        this.subTasks = [];

        /** @type {PollableTask} */
        this.parentTask = null;

        /** @type {Number} */
        this.timeout = 0;

        /** @type {User} */
        this.createdByUser = null;

        /** @type {Boolean} */
        this.isAllFinished = false;
    }

    /**
     * Convert JSON User object
     *
     * @param {Object} json
     * @return {PollableTask}
     */
    static toPollableTask(json) {
        let result = null;

        if (json) {
            result = new PollableTask();

            result.id = json.id;

            result.name = json.name;

            result.finishedDate = new Date(json.finishedDate);

            result.message = json.message;

            if (json.errorMessage) {
                result.errorMessage = json.errorMessage;
                result.errorStack = json.errorStack;
            }

            result.expectedSubTaskNumber = json.expectedSubTaskNumber;
            result.subTasks = PollableTask.toPollableTasks(json.subTasks);
            result.parentTask = PollableTask.toPollableTask(json.parentTask);
            result.timeout = json.timeout;
            result.createdByUser = User.toUser(json.createdByUser);

            // NOTE PollableTask.isAllFinished json property is just "allFinished"
            result.isAllFinished = json.allFinished;
        }

        return result;
    }

    /**
     * @param {Object[]} jsons
     * @return {PollableTask[]}
     */
    static toPollableTasks(jsons) {
        let results = null;
        if (jsons) {
            results = [];
            for (const json of jsons) {
                results.push(PollableTask.toPollableTask(json));
            }
        }
        return results;
    }
}
